servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,681 @@
+                        terminalNotification.terminate(target, de);
+                    }
+                } else {
+                    terminalNotification.terminate(target);
+                }
+                return false;
+            }
+            @SuppressWarnings("unchecked")
+            final R rItem = item == NULL_TOKEN ? null : (R) item;
+            target.onNext(rItem);
+            return true;
+        }
+
+        private boolean acquireTryRequestMoreLock() {
+            return acquirePendingLock(tryRequestMoreLockUpdater, this);
+        }
+
+        private boolean releaseTryRequestMoreLock() {
+            return releasePendingLock(tryRequestMoreLockUpdater, this);
+        }
+
+        private void tryRequestMore(FlatMapPublisherSubscriber<T, R> subscriber) {
+            if (acquireTryRequestMoreLock()) {
+                try {
+                    final long availableRequestN = pendingDemandUpdater.getAndSet(this, 0);
+                    if (availableRequestN != 0) {
+                        final int consumedRequestN = subscriber.requestFromUpstream(
+                                (int) min(source.maxMappedDemand, availableRequestN));
+                        assert availableRequestN >= consumedRequestN;
+                        if (consumedRequestN != availableRequestN) {
+                            giveBackRequestN(availableRequestN - consumedRequestN);
+                        }
+                    }
+                } finally {
+                    if (!releaseTryRequestMoreLock()) {
+                        tryRequestMore();
+                    }
+                }
+            }
+        }
+
+        private void tryRequestMore() {
+            Throwable delayedCause = null;
+            for (;;) {
+                if (!acquireTryRequestMoreLock()) {
+                    break;
+                }
+
+                try {
+                    final long availableRequestN = pendingDemandUpdater.getAndSet(this, 0);
+                    if (availableRequestN != 0) {
+                        long remainingRequestN = availableRequestN;
+                        Iterator<FlatMapPublisherSubscriber<T, R>> itr = subscribers.iterator();
+                        while (itr.hasNext() && remainingRequestN > 0) {
+                            remainingRequestN -= itr.next().requestFromUpstream(
+                                    (int) Math.min(source.maxMappedDemand, remainingRequestN));
+                        }
+
+                        assert availableRequestN >= remainingRequestN && remainingRequestN >= 0;
+                        if (remainingRequestN > 0) {
+                            giveBackRequestN(remainingRequestN);
+                        }
+                    }
+                } catch (Throwable cause) {
+                    if (delayedCause == null) {
+                        delayedCause = cause;
+                    }
+                }
+                if (releaseTryRequestMoreLock()) {
```

Comment:
same comment as above... I think we want to call this method in a finally block.

Reply:
+1 I'll put in a finally